Dockerfile

```diff
@@ -1,16 +1,27 @@
 # ============ Build Stage ============
-FROM rust:1.74-bookworm as build
-
+FROM lukemathwalker/cargo-chef:latest-rust-1 AS chef
 WORKDIR /typst-bot
 
+FROM chef AS planner
+
 # Compilation requires only the source code.
 COPY Cargo.toml Cargo.lock ./
 COPY protocol protocol
 COPY worker worker
 COPY bot bot
+RUN cargo chef prepare --recipe-path recipe.json
 
-RUN cargo build --release --all --config git-fetch-with-cli=true
+FROM chef AS builder
 
+COPY --from=planner /typst-bot/recipe.json recipe.json
+# Caching layer
+RUN cargo chef cook --release --workspace --recipe-path recipe.json
+# Compilation requires only the source code.
+COPY Cargo.toml Cargo.lock ./
+COPY protocol protocol
+COPY worker worker
+COPY bot bot
```

Comment:
Since Docker creates a new layer per each `COPY` (even though a long time ago I read that it sometimes can auto-merge them) and you are using the same operation multiple times, it will create an inefficient Docker image. Following [the example](https://github.com/LukeMathWalker/cargo-chef/blob/b468537839bfc7c23d744b85d7a5090954626550/README.md?plain=1#L149), it should be a single `COPY . .`. To not bloat the context, you can either try combining all these lines into one or a `.dockerignore` file can be added and everything unused by the `COPY . .` command will have to go into that file.

Lines [8–11](https://github.com/mattfbacon/typst-bot/pull/23/files#diff-dd2c0eb6ea5cfc6c4bd4eac30934e2d5746747af48fef6da689e85b752f39557R8-R11) can be left unchanged, or you can optimize them as well. :)

You can just replace these with `COPY . .` and we can then open another PR perfecting this repo after this gets merged.

Reply:
I assume the best-practice approach is use '.dockerignore', since all other projects I have seen do a 'COPY . .'. Extra files shouldn't break cargo-chef, anyway.

Reply:
> I assume the best-practice approach is use '.dockerignore', since all other projects I have seen do a 'COPY . .'.

Yes, I think so too.

> Extra files shouldn't break cargo-chef, anyway.

They won't but the context will increase otherwise. The `target` dir will then also be transferred to the context. Not good.

I think `.dockerignore` also supports reverse patterns so that instead of listing everything that should be excluded, you can instead list everything that should be included. I'm pretty sure that `cargo-chef` only neads `Cargo.toml`, `Cargo.lock`, `lib.rs`, and `main.rs`.

Reply:
Except... if `./fonts` is added to `.dockerignore`, will it be possible to `COPY` it at the end?

Reply:
...no, it looks like it won't be possible. I thought it only applies to the `.`, but it seems it applies to everything. So the fonts directory is also should be excluded from the ignore list.
https://docs.docker.com/build/building/context/#negating-matches

Reply:
I'll do a workaround with `COPY --exclude [...] . .`

Reply:
It's not a stable feature. Not sure if everyone can access it (from a non-latest version) or will it be (re)moved after some time.
https://docs.docker.com/reference/dockerfile/#copy---exclude

Reply:
I can't access it 😭

Reply:
Told ya. :D

Reply:
It's better to have extra layers than revoking the cache because the fonts changed. Marking as resolved.

Reply:
Well, this is annoying. With the current `COPY`'s syntax and the fact that there is an additional `COPY` at the end, there isn't much we can do in terms of eliminating the multiple adjacent `COPY` commands...
We can add a tiny bit of preprocessing like `mkdir -p dirs && cp -r protocol worker bot dirs` and then write `COPY dirs Cargo.toml Cargo.lock ./`after which run `rm -rf dirs`. But this adds more hassle than it solves, maybe. So yeah, I guess we will keep it like that for now.

https://github.com/docker/roadmap/issues/164#issuecomment-2184107422

Reply:
Yeah, what I would do is to organise the project structure better, but thats out of scope.

Reply:
I'm interested, what is your suggestion?

Reply:
Put all the source code in one `./src` folder, or whatever the convention is for a workspace

Reply:
I think the closest to a convention would be a `crates` dir. I'll do that.

Reply:
OK rebase on head

Reply:
Cool, but this would only reduce the number of `COPY` commands to 2: `COPY Cargo.toml Cargo.lock ./`and `COPY crates crates`, which is of course better than 4, but still isn't 1. Since it's used twice, it would reduce the number of lines used (and layers created) from 8 to 4.

You know what? The non-last stages won't create any layers in the final image, so for the end image this doesn't matter. For the building...also not really. Because it's pretty easy (fast) to copy all the source code in parts, the only downside is that the non-last stages still add some cache/layers (to the host) so it would increase the cache size. I don't think there are any other issues with it, other than that there are multiple lines of code that in theory can be made into one (`cp -r Cargo.toml Cargo.lock crates dest`), i.e., it's a bit verbose (the `Dockerfile`).

Reply:
@Andrew15-5 Nope it can't be done that way. That would copy the contents of the folders (shoddy design).

@mattfbacon Can cargo.toml and cargo.lock be moved to the base of `crates`?

Reply:
> it can't be done that way.

What exactly are you referring to?

Reply:
`cp -r Cargo.toml Cargo.lock crates dest` but with Dockerfile syntax

Reply:
I feel like this isn't really our problem; Docker should be capable of combining multiple `COPY`s into one layer.

> Can cargo.toml and cargo.lock be moved to the base of `crates`?

No, I don't think that's correct. Our current structure mirrors https://github.com/typst/typst

Reply:
> `cp -r Cargo.toml Cargo.lock crates dest` but with Dockerfile syntax

That's the whole point of this discussion. I gave an example of what _in theory_ it should look like (just one line instead of many). I know that it can't be done that way. That's why I created this comment: https://github.com/docker/roadmap/issues/164#issuecomment-2184107422.